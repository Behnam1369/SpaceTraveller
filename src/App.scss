body {
  header {
    padding: 30px;

    .logo {
      width: 40px;
      height: 40px;
    }

    nav {
      float: right;
    }

    nav > a {
      text-decoration: none;
      color: blue;
      margin: 20px;
    }

    nav > a.active {
      text-decoration: underline;
      color: red;
    }
  }

  .missionList {
    width: 90%;
    margin: auto;
    border-collapse: collapse;

    thead {
      th {
        border: solid 1px silver;
        font-family: Helvetica, sans-serif;
        padding: 5px;
      }
    }

    tbody tr:nth-child(odd) {
      background-color: whitesmoke;
    }

    td {
      border: solid 1px silver;
      font-family: Helvetica, sans-serif;
      padding: 5px;

      span {
        background-color: gray;
        color: white;
        white-space: nowrap;
        padding: 5px;
        font-family: Helvetica, sans-serif;
        font-size: 14px;
        border-radius: 3px;
      }

      span.joined {
        background-color: darkcyan;
      }
    }

    td:nth-child(1) {
      font-weight: bold;
      text-align: center;
    }

    td:nth-child(2) {
      font-family: Arial, sans-serif;
      text-align: left;

      button {
        background-color: transparent;
        border: none;
        color: blue;
        cursor: pointer;
      }
    }

    td:nth-child(4) button {
      background-color: white;
      color: gray;
      border: solid 1px gray;
      white-space: nowrap;
      padding: 5px;
      font-family: Helvetica, sans-serif;
      font-size: 14px;
      cursor: pointer;
      border-radius: 4px;
    }

    td:nth-child(4) button.joined {
      color: red;
      border-color: red;
    }
  }

  .profile {
    margin: 7%;

    h1 {
      font-family: Helvetica, sans-serif;
    }

    h2 {
      font-family: Helvetica, sans-serif;
    }

    ul {
      width: 50%;
      margin: 0;
      padding: 0;

      li {
        list-style: none;
        margin: 0;
        padding: 20px;
        border: solid 1px lightgray;
      }
    }
  }
}
