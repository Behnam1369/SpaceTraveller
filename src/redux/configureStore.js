import { combineReducers, createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import missionReducer from './missions';
import { rocketsReducer } from './rockets';

const rootReducer = combineReducers({
  rockets: rocketsReducer,
  missions: missionReducer,
});

const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;
