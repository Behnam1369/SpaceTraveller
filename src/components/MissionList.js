import { React, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Mission from './Mission';
import { getMissions } from '../redux/missions';

function MissionList() {
  const {
    loading, missions, failureAPI,
  } = useSelector((state) => state.missions);

  const dispatch = useDispatch();
  useEffect(() => {
    if (missions === null) {
      dispatch(getMissions());
    }
  }, []);

  let missionList = '';
  if (loading) {
    missionList = <h3>Loading...</h3>;
  } else if (failureAPI) {
    missionList = <h3>OOPS! SOMETHING WENT WRONG. PLEASE TRY AGAIN LATER.</h3>;
  } else if (missions.length === 0) {
    missionList = <h3>There is no book in the list yet.</h3>;
  } else {
    missionList = missions.map((el) => (
      <Mission
        key={el.mission_id}
        description={el.description}
        name={el.mission_name}
      />
    ));
  }
  return (
    <div className="missionList">
      {missionList}
    </div>
  );
}

export default MissionList;
