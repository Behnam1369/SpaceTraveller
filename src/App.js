import React from 'react';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import Header from './components/Header';
import MyProfile from './components/MyProfile';
import './App.scss';
import MissionList from './components/MissionList';

function App() {
  return (
    <div className="App">
      <BrowserRouter>
        <Header />
        <Routes>
          <Route path="/MyProfile" element={<MyProfile />} />
          <Route path="/Missions" element={<MissionList />} />
        </Routes>
      </BrowserRouter>
    </div>
  );
}

export default App;
