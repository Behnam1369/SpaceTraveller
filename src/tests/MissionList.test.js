import TestRenderer from 'react-test-renderer';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import MissionList from '../components/MissionList';
import '@testing-library/jest-dom';
import configureStore from '../redux/configureStore';
import App from '../App';

describe('Unit Testing Mission', () => {
  it('Test Mission component snapshot', () => {
    const testRenderer = TestRenderer.create(
      <Provider store={configureStore}>
        <MissionList />
      </Provider>,
    ).toJSON();
    expect(testRenderer).toMatchSnapshot();
  });
});

describe('Testing Mission in Application', () => {
  it('Test join functionality', () => {
    const missions = [
      {
        mission_name: 'TEST MISSION',
        mission_id: 'key1',
        description: 'TEST DESCRIPTION',
        reserved: false,
      },
    ];

    configureStore.dispatch({ type: 'space-traveller/missions/GET MISSIONS', missions });
    configureStore.dispatch({ type: 'book-store/books/SET LOADING', loading: false });
    render(
      <Provider store={configureStore}>
        <App />
      </Provider>,
    );
    fireEvent.click(screen.getByText('Missions'));
    waitFor(() => screen.getByText('Join Mission')[0]);
    fireEvent.click(screen.getByText('Join Mission'));
    expect(screen.getByTestId('badge')).toHaveTextContent('Active Member');
  });

  it('Test leave functionality', () => {
    const missions = [
      {
        mission_name: 'TEST MISSION',
        mission_id: 'key1',
        description: 'TEST DESCRIPTION',
        reserved: false,
        joined: true,
      },
    ];

    configureStore.dispatch({ type: 'space-traveller/missions/GET MISSIONS', missions });
    configureStore.dispatch({ type: 'book-store/books/SET LOADING', loading: false });
    render(
      <Provider store={configureStore}>
        <App />
      </Provider>,
    );
    fireEvent.click(screen.getByText('Missions'));
    waitFor(() => screen.getByText('Leave Mission')[0]);
    fireEvent.click(screen.getByText('Leave Mission'));
    expect(screen.getByTestId('badge')).toHaveTextContent('NOT A MEMBER');
  });
});
