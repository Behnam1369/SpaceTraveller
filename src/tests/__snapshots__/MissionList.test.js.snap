// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit Testing Mission Test Mission component snapshot 1`] = `
<table
  className="missionList"
>
  <thead>
    <tr>
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th>
         
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={4}
      >
        Loading...
      </td>
    </tr>
  </tbody>
</table>
`;
